src/installer/src/main.rs

```diff
@@ -0,0 +1,232 @@
+    println!("Please select (0-{}): ", ghidra_locations.len() - 1);
+
+    let mut choice = String::new();
+
+    std::io::stdin().read_line(&mut choice)?;
+
+    match choice.trim().parse::<usize>() {
+        Ok(i) => {
+            if i < ghidra_locations.len() {
+                Ok(ghidra_locations[i].clone())
+            } else {
+                Err(Error::new(ErrorKind::InvalidInput, "invalid user input").into())
+            }
+        }
+        Err(error) => Err(error.into()),
+    }
+}
+
+/// Creates ghidra.json for a Ghidra location at provided locaton.
+fn create_ghidra_json(location: &Path, ghidra_location: PathBuf) -> Result<()> {
+    let conf = GhidraConfig {
+        ghidra_path: ghidra_location,
+    };
+
+    std::fs::write(location.join("ghidra.json"), serde_json::to_string(&conf)?)?;
+    Ok(())
+}
+
+/// Runs Cargo install to install cwe_checker.
+fn install_cwe_checker() -> Result<()> {
+    std::process::Command::new("cargo")
+        .args(["install", "--path", "src/caller", "--locked"])
+        .status()?;
+    Ok(())
+}
+
+// Recursive copy of files and directories.
+fn copy_dir_all(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> Result<()> {
+    fs::create_dir_all(&dst)?;
+    for entry in fs::read_dir(src)? {
+        let entry = entry?;
+        let ty = entry.file_type()?;
+        if ty.is_dir() {
+            copy_dir_all(entry.path(), dst.as_ref().join(entry.file_name()))?;
+        } else {
+            fs::copy(entry.path(), dst.as_ref().join(entry.file_name()))?;
+        }
+    }
+    Ok(())
+}
+
+// Copy src/ghidra to provided location.
+fn copy_ghidra_plugin(target: &Path) -> Result<()> {
+    let target = &target;
+    let mut source = env::current_dir()?;
+    source.push("src/ghidra");
+
+    copy_dir_all(source, target.join("ghidra"))?;
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    let args: Vec<String> = env::args().collect();
+    let cwe_checker_conf_dir = ProjectDirs::from("", "", "cwe_checker").unwrap();
+    std::fs::create_dir_all(cwe_checker_conf_dir.config_dir())?;
+
+    println!("creating config.json...");
+    copy_config_json(cwe_checker_conf_dir.config_dir())?;
+
+    println!("create ghidra.json...");
+    if args.len() == 2 {
+        create_ghidra_json(cwe_checker_conf_dir.config_dir(), PathBuf::from(&args[1]))?;
+    } else if cwe_checker_conf_dir
+        .config_dir()
+        .join("ghidra.json")
+        .exists()
+    {
+        println!(
+            "found ghidra.json at {}, keeping it.",
+            cwe_checker_conf_dir.config_dir().display()
+        );
+    } else {
+        println!("searching for ghidra...");
+        match find_ghidra() {
+            Ok(ghidra_location) => {
+                create_ghidra_json(cwe_checker_conf_dir.config_dir(), ghidra_location)?;
+            }
+            Err(err) => return Err(err),
+        }
+    }
+
+    println!("installing CWE-Checker...");
+    install_cwe_checker()?;
```

Comment:
We should make sure that we do not needlessly copy configuration files if the compilation step fails. So we should do the compilation step before any config file creation (but after selection of the ghidra path).